benchmarks/src/main/java/org/apache/druid/benchmark/query/CachingClusteredClientBenchmark.java

```diff
@@ -0,0 +1,569 @@
+    );
+  }
+
+  private static GroupByQueryRunnerFactory makeGroupByQueryRunnerFactory(
+      final ObjectMapper mapper,
+      final GroupByQueryConfig config,
+      final DruidProcessingConfig processingConfig
+  )
+  {
+    final Supplier<GroupByQueryConfig> configSupplier = Suppliers.ofInstance(config);
+    final Supplier<ByteBuffer> bufferSupplier =
+        () -> ByteBuffer.allocateDirect(processingConfig.intermediateComputeSizeBytes());
+
+    final NonBlockingPool<ByteBuffer> bufferPool = new StupidPool<>(
+        "GroupByQueryEngine-bufferPool",
+        bufferSupplier
+    );
+    final BlockingPool<ByteBuffer> mergeBufferPool = new DefaultBlockingPool<>(
+        bufferSupplier,
+        processingConfig.getNumMergeBuffers()
+    );
+    final GroupByStrategySelector strategySelector = new GroupByStrategySelector(
+        configSupplier,
+        new GroupByStrategyV1(
+            configSupplier,
+            new GroupByQueryEngine(configSupplier, bufferPool),
+            QueryRunnerTestHelper.NOOP_QUERYWATCHER,
+            bufferPool
+        ),
+        new GroupByStrategyV2(
+            processingConfig,
+            configSupplier,
+            bufferPool,
+            mergeBufferPool,
+            mapper,
+            QueryRunnerTestHelper.NOOP_QUERYWATCHER
+        )
+    );
+    final GroupByQueryQueryToolChest toolChest = new GroupByQueryQueryToolChest(
+        strategySelector,
+        QueryRunnerTestHelper.sameThreadIntervalChunkingQueryRunnerDecorator()
+    );
+    return new GroupByQueryRunnerFactory(
+        strategySelector,
+        toolChest
+    );
+  }
+
+  @TearDown(Level.Trial)
+  public void tearDown() throws IOException
+  {
+    closer.close();
+    processingPool.shutdown();
+  }
+
+  @Benchmark
+  @BenchmarkMode(Mode.AverageTime)
+  @OutputTimeUnit(TimeUnit.MICROSECONDS)
+  public void timeseriesQuery(Blackhole blackhole)
+  {
+    query = Druids.newTimeseriesQueryBuilder()
+                  .dataSource(DATA_SOURCE)
+                  .intervals(basicSchemaIntervalSpec)
+                  .aggregators(new LongSumAggregatorFactory("sumLongSequential", "sumLongSequential"))
+                  .granularity(Granularity.fromString(queryGranularity))
+                  .build();
+
+    final List<Result<TimeseriesResultValue>> results = runQuery();
+
+    for (Result<TimeseriesResultValue> result : results) {
+      blackhole.consume(result);
+    }
+  }
+
+  @Benchmark
+  @BenchmarkMode(Mode.AverageTime)
+  @OutputTimeUnit(TimeUnit.MICROSECONDS)
+  public void topNQuery(Blackhole blackhole)
+  {
+    query = new TopNQueryBuilder()
+        .dataSource(DATA_SOURCE)
+        .intervals(basicSchemaIntervalSpec)
+        .dimension(new DefaultDimensionSpec("dimUniform", null))
+        .aggregators(new LongSumAggregatorFactory("sumLongSequential", "sumLongSequential"))
+        .granularity(Granularity.fromString(queryGranularity))
+        .metric("sumLongSequential")
+        .threshold(20480)
```

Comment:
I talked to @jihoonson (who wrote the benchmark originally) and we don't believe the number is actually significant, it was just an arbitrarily large number to make the result meaningful. I changed it to 10000 and have added a comment about it. I think it likely this will maybe be moved into a benchmark parameter in the future to compare merges for different thresholds, but I will save that until I actually have different merges to compare.

Reply:
I think I set it to 20480 in #6629, but don't remember where this number came from. I think `threshold` could be any number if it's not too small so that parallel merge can help with merging results.